From 4312377de576a113eefdf6874a44eea2ac2f3e05 Mon Sep 17 00:00:00 2001
From: David Edmundson <davidedmundson@kde.org>
Date: Wed, 8 Apr 2020 18:19:00 +0100
Subject: [PATCH 04/18] Client: Reset wl_egl_window before wl_surface

Semantically we want things in this order as wl_egl_window_create takes
a reference to the wl_surface. Teardown typically is the opposite of
construction.

This brings us in line with GTK which also tears down in this order.

This fixes a repeatable crash when using Nvidia's EGLStreams and closing
a window.

Change-Id: I356dfbba9a7315717c8ecdc7023ca7b047622697
Reviewed-by: Frederik Gladhorn <gladhorn@kde.org>
Reviewed-by: Johan Helsing <johanhelsing@gmail.com>
(cherry picked from commit 4bd13402f0293e85b8dfdf92254e250ac28094c7)

Conflicts:
	src/client/qwaylandwindow.cpp
	src/hardwareintegration/client/wayland-egl/qwaylandeglwindow.cpp

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 src/client/qwaylandwindow.cpp                              | 3 +++
 .../client/wayland-egl/qwaylandeglwindow.cpp               | 7 -------
 .../client/wayland-egl/qwaylandeglwindow.h                 | 1 -
 3 files changed, 3 insertions(+), 8 deletions(-)

diff --git a/src/client/qwaylandwindow.cpp b/src/client/qwaylandwindow.cpp
index e7a01a54..43f85d57 100644
--- a/src/client/qwaylandwindow.cpp
+++ b/src/client/qwaylandwindow.cpp
@@ -248,6 +248,9 @@ void QWaylandWindow::reset()
     mShellSurface = 0;
     delete mSubSurfaceWindow;
     mSubSurfaceWindow = 0;
+
+    invalidateSurface();
+
     if (isInitialized())
         destroy();
 
diff --git a/src/hardwareintegration/client/wayland-egl/qwaylandeglwindow.cpp b/src/hardwareintegration/client/wayland-egl/qwaylandeglwindow.cpp
index 2a4ec332..1ae52d51 100644
--- a/src/hardwareintegration/client/wayland-egl/qwaylandeglwindow.cpp
+++ b/src/hardwareintegration/client/wayland-egl/qwaylandeglwindow.cpp
@@ -158,13 +158,6 @@ QSurfaceFormat QWaylandEglWindow::format() const
     return m_format;
 }
 
-void QWaylandEglWindow::setVisible(bool visible)
-{
-    QWaylandWindow::setVisible(visible);
-    if (!visible)
-        QMetaObject::invokeMethod(this, "doInvalidateSurface", Qt::QueuedConnection);
-}
-
 void QWaylandEglWindow::doInvalidateSurface()
 {
     if (!window()->isVisible())
diff --git a/src/hardwareintegration/client/wayland-egl/qwaylandeglwindow.h b/src/hardwareintegration/client/wayland-egl/qwaylandeglwindow.h
index e471a8f7..b45e7b07 100644
--- a/src/hardwareintegration/client/wayland-egl/qwaylandeglwindow.h
+++ b/src/hardwareintegration/client/wayland-egl/qwaylandeglwindow.h
@@ -74,7 +74,6 @@ public:
     void bindContentFBO();
 
     void invalidateSurface() override;
-    void setVisible(bool visible) override;
 
 private Q_SLOTS:
     void doInvalidateSurface();
-- 
2.20.1

